From fbeccbb2c4e5b1fc3eb8c059f9f1b993a869bc02 Mon Sep 17 00:00:00 2001
From: Evert Jans <evert@mintconsulting.be>
Date: Tue, 20 Jan 2015 22:13:01 +0100
Subject: [PATCH] added decent makefile

---
 Makefile | 28 ++++++++++++++++++++++++++--
 1 file changed, 26 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index 6cf661e..ccbf5eb 100644
--- a/Makefile
+++ b/Makefile
@@ -1,2 +1,26 @@
-zpaq: zpaq.cpp libzpaq.cpp libzpaq.h
-	g++ -O3 -march=native -Dunix zpaq.cpp libzpaq.cpp -pthread -o zpaq
+CCFLAGS= -O3 -c -DNDEBUG
+PPFLAGS= -O3 -march=native -Dunix zpaq.cpp libzpaq.cpp -pthread
+prefix=/usr/local
+BIN=${prefix}/bin
+INC=${prefix}/include
+
+all: zpaq
+
+zpaq: zpaq.cpp libzpaq.cpp libzpaq.h
+	g++ $(PPFLAGS) -o $@
+
+install: all
+	@echo installing executable file to ${BIN}
+	@mkdir -p ${BIN}
+	@cp -f zpaq ${BIN}
+	@chmod 755 ${BIN}/zpaq
+
+uninstall:
+	@echo removing executable file from ${BIN}
+	@rm -f ${BIN}/zpaq
+
+clean:
+	rm -vf zpaq
-- 
1.9.1

