From e84e2cd48b0e07a399f3df722f649cf587c876be Mon Sep 17 00:00:00 2001
From: Evert Jans <evert@sonuwe.com>
Date: Thu, 29 Oct 2015 18:17:22 +0100
Subject: [PATCH] evert fix

---
 Makefile | 23 ++++++++++++++++++++++-
 1 file changed, 22 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 6cf661e..1ae01ac 100644
--- a/Makefile
+++ b/Makefile
@@ -1,2 +1,23 @@
+CCFLAGS= -O3 -c -DNDEBUG
+PPFLAGS= -O3 -march=native -Dunix zpaq.cpp libzpaq.cpp -pthread
+prefix=/usr/local
+BIN=${prefix}/bin
+INC=${prefix}/include
+
+all: zpaq
+
 zpaq: zpaq.cpp libzpaq.cpp libzpaq.h
-	g++ -O3 -march=native -Dunix zpaq.cpp libzpaq.cpp -pthread -o zpaq
+	g++ $(PPFLAGS) -o $@
+
+install: all
+	@echo installing executable file to ${BIN}
+	@mkdir -p ${BIN}
+	@cp -f zpaq ${BIN}
+	@chmod 755 ${BIN}/zpaq
+
+uninstall:
+	@echo removing executable file from ${BIN}
+	@rm -f ${BIN}/zpaq
+
+clean:
+	rm -vf zpaq
-- 
2.5.0

